### ISSUE TITLE
Bug: Wrong new line (\n) in some cases with the option shouldPreverseNewLines = true

### ISSUE THREAD

**Issue by FrancisRoc** [facebook/lexical#6418]
Lexical version: 0.16.1

## Steps To Reproduce
It seems that the update to add `shouldPreserveNewLines` on `$convertToMarkdownString` and `convertFromMarkdownString` results in wrong output string in some cases ?

If you keep this new option to false and enter some text > press 'Enter' > enter some other text
it results in the right thing: `text\n\nsome other text`. Which will also properly display in markdown (2 lines of test)
![image](https://github.com/user-attachments/assets/31012105-0efe-44d4-8d79-92b964c237de)

BUT, if you but those options to true you end up with `text/nsome other text` which is wrong markdown (it will be 1 line of text). 
![image](https://github.com/user-attachments/assets/d9ca1dbc-8898-41a3-9a5a-97ed12083bfc)

The second comment in those changes seems accurate but I don't think that it's what the logic accomplishes 🤔 
![image](https://github.com/user-attachments/assets/7d5361a9-6a54-4ad2-8333-a1ef595d0d04)
Maybe I am missing something but when I input that in the playground I end up with `"hello\n\nhi\nworld"`

Maybe it's more a question than a bug report :o 

## The current behavior
`shouldPreverseNewLines=true` on markdown conversion results in single `\n` in between text paragraphs but when it's set to false it results in proper `\n\n`

## The expected behavior
If we ignore empty lines to simplify, `shouldPreverseNewLines=true` should not impact the number of `\n` in between paragraph nodes (should be 2)

## Impact of fix

Breaks the output that will not work with markdown. The issue is that the behaviour of `shouldPreserveNewLine=false` seems to also have change. An empty line won't add only 1 extra `\n` (before the change it would add 2) so if we had some patch/workaround to deal with empty lines in markdown then the new version breaks this

**Comment by potatowagon** [comment]
im not sure if i understand this bug report

> BUT, if you but those options to true you end up with text/nsome other text which is wrong markdown (it will be 1 line of text).

im not able to reproduce the above ^, see recording, text does not appear on one line

https://github.com/user-attachments/assets/cbf778ac-46fe-4332-b211-6f82cd42b82a

**Comment by FrancisRoc** [comment]
Hey @potatowagon , sorry for the late reply. I am not sure if it's an actual bug or the lack of standards for markdown strings. But if you copy the text you have in your video when it's rendered in markdown and paste it in JSON.strignify you will see that there is only 1 `\n`.  
Then, in some editors (maybe not all which is why maybe it's not a bug) if you print the same text of your markdown (text\nsome other text) it will display all on a single line.
Example of libraries doing so is react-markdown and also this react-md-editor: https://uiwjs.github.io/react-md-editor/

Now maybe it's just cause of my use case that it makes this harder to work with but it seems that when you then use that markdown output to send emails, a single `\n` will not make a new line neither. Talking about Sendgrid more specifically.

I hope this makes it clearer, feels like I am doing a bad job at explaining 😅

**Comment by potatowagon** [comment]
hi @FrancisRoc , shouldPreverseNewLines = true is a feature that deviates from markdown spec, since markdown spec does not preserve new lines. perhaps try setting shouldPreverseNewLines = false if following markdown spec is what youre looking for

**Comment by FrancisRoc** [comment]
Thanks for quick replis @potatowagon , I really appreciate it!
Yeah I think I have a weird case where I want to follow markdown except for newlines (I want to allow empty lines). In the end I can't really ask for you guys to accomodate my use cases.

The only thing I would say is that maybe it's less confusing if we end up with 2 `\n` weather `shouldPreverseNewLines` is true or false. Right now we already discussed the case where it's true and we saw in your screenrecord that it works in the playground and we also discussed that it has 1 `\n` (`text\nsome other text`)

But now if I use `shouldPreserveNewLines=false` and I display it in markdown on the lexical playground, the result is different (1 enter will result in 2 `\n` in markdown which leaves an empty line vs. no empty lines in the editor - see screenshots). I would think that when there is no empty lines the result should be the same for both.

**preverve line breaks false** result in markdown that is not the same than the editor text
![image](https://github.com/user-attachments/assets/d97ffd79-ac01-44af-893e-5f3f1a1f96bb)
![image](https://github.com/user-attachments/assets/a5c56294-9cb7-4f8d-92bf-651b6b7ad01a)
